app/src/main/java/com/rkm/tasky/repository/implementation/TaskyEventRepositoryImpl.kt

```diff
@@ -0,0 +1,115 @@
+package com.rkm.tasky.repository.implementation
+
+import com.rkm.tasky.database.dao.AttendeeDao
+import com.rkm.tasky.database.dao.EventDao
+import com.rkm.tasky.database.dao.PhotoDao
+import com.rkm.tasky.database.dao.SyncDao
+import com.rkm.tasky.database.error.DatabaseError
+import com.rkm.tasky.database.model.SyncEntity
+import com.rkm.tasky.database.model.SyncItemType
+import com.rkm.tasky.database.model.SyncUserAction
+import com.rkm.tasky.di.IoDispatcher
+import com.rkm.tasky.feature.event.model.Event
+import com.rkm.tasky.network.datasource.TaskyEventRemoteDataSource
+import com.rkm.tasky.network.util.NetworkError
+import com.rkm.tasky.network.util.asMultiPartBody
+import com.rkm.tasky.network.util.safeCall
+import com.rkm.tasky.repository.abstraction.TaskyEventRepository
+import com.rkm.tasky.repository.mapper.asAttendeeEntity
+import com.rkm.tasky.repository.mapper.asCreateEventRequest
+import com.rkm.tasky.repository.mapper.asEventEntity
+import com.rkm.tasky.repository.mapper.asPhotoEntity
+import com.rkm.tasky.repository.mapper.asUpdateEventRequest
+import com.rkm.tasky.repository.mapper.toEvent
+import com.rkm.tasky.util.result.EmptyResult
+import com.rkm.tasky.util.result.Result
+import com.rkm.tasky.util.result.asEmptyDataResult
+import com.rkm.tasky.util.result.onFailure
+import com.rkm.tasky.util.result.onSuccess
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class TaskyEventRepositoryImpl @Inject constructor(
+    private val remoteDataSource: TaskyEventRemoteDataSource,
+    private val localDataSource: EventDao,
+    private val attendeeLocalDataSource: AttendeeDao,
+    private val photoLocalDataSource: PhotoDao,
+    private val syncDataSource: SyncDao,
+    @IoDispatcher private val dispatcher: CoroutineDispatcher
+) : TaskyEventRepository {
+    override suspend fun getEvent(id: String): Result<Event, DatabaseError.ItemError> =
+        withContext(dispatcher) {
+            return@withContext when (val results = localDataSource.getAllEventDetails(id)) {
+                null -> Result.Error(DatabaseError.ItemError.ITEM_DOES_NOT_EXIST)
+                else -> Result.Success(results.toEvent())
+            }
+        }
+
+    override suspend fun createEvent(event: Event): EmptyResult<NetworkError.APIError> =
+        withContext(dispatcher) {
+            val result = safeCall {
+                remoteDataSource.createEvent(
+                    event.asCreateEventRequest(),
+                    event.photosToUpload.asMultiPartBody()
+                )
+            }
+            result.onSuccess {
+                localDataSource.upsertEvent(it.asEventEntity())
+                attendeeLocalDataSource.upsertAttendees(it.attendees.map { attendee -> attendee.asAttendeeEntity() })
+                photoLocalDataSource.upsertPhotos(it.photos.map { photo -> photo.asPhotoEntity(it.id) })
```

Comment:
Even though very unlikely, you'll need to take some care when launching two such suspend functions one after another when either both have to be executed or none. Imagine, you'd finish the first call of `taskyApi.updateTask(...)` and then before the code reaches the next suspend function `taskDao.insertModifiedTaskId(...)`, the coroutine is cancelled. This could happen when this suspend fun is launched in `viewModelScope` for example and the user navigates back. In that case the line to update the local DB won't be executed, however the API was already notified about that change. So the app thinks the task is not pending even though it is (and this could lead to invalid states, since your task will never be synced until it's edited again).

The fix is to make sure that the second suspend function is not cancellable. The best way to do this is with a custom application wide coroutine scope, which lives as long as your app's process:
```
class MyApp: Application() {
    val applicationScope = CoroutineScope(SupervisorJob())
    
    // ...
}
```
You can then provide an instance of this scope in your Dagger module:
```
fun provideApplicationScope(app: Application): CoroutineScope {
    return (app as MyApp).applicationScope
}
```
Then, inject it into the class you need it in:
```
class MyRepository(
    private val applicationScope: CoroutineScope,
    ...
)
```
And make sure the second suspend function is executed in this new scope which won't be cancelled:
```
applicationScope.launch {
    // DB call
}.join()
```
This way, if the coroutine reached this point and wasn't cancelled, we make sure that it also won't be cancelled during this second call. The `join()` makes sure to still suspend the outer suspend function until the DB call finished.